request.go

```diff
@@ -90,7 +93,9 @@ func (req *Request) DefaultHeaders(config *Config, userAgent string) (err error)
 
 		req.Header.Add(HeaderAuth, url.QueryEscape("type=master&ver=1.0&sig="+sign))
 	} else if config.ServicePrincipal != nil {
-		err := config.ServicePrincipal.EnsureFresh()
+		ctx, cancel := context.WithTimeout(context.Background(), ServicePrincipalRefreshTimeout)
```

Comment:
Don't we want to use here the request context instead of background? `req.Context()`